### PR TITLE
[16364] Switch WorkPackage show view over to AngularJS (aka fullscreen view)

### PR DESCRIPTION
**Pull request by timhabermaas**
https://community.openproject.org/work_packages/16364
- [ ] fix specs
- [ ] clean up directives/code duplication
- [ ] being in split view and clicking on a work package does not update the right pane
- [x] "Watchers" does not show watchers and one can't add watchers. see split view for correct implementation
#### Review Feedback

Feedback von @lindenthal und @basiszwo
- [x] Inline editing of WP title missing
- [x] project context is being lost when switching to fullscreen view `/project/seeded_project/work_packages/1` vs `/work_packages/1`
- [x] "more" Button with actions in dropdown is missing [1]
- [x] "Edit" Button should link to old edit view (same as in split view) [1]
- [x] cog icon on top right does not make sense in fullscreen view as it provides list settings
  - replace with "more" Button?
- [x] missing full screen icon (already included in icon set)
- [x] scrollbar should not be visible when there's nothing to scroll (@rebellium)
- [x] Validate/Fix the actions which should still work (e.g. watching a work package); see OpenProject's work package
- [x] Title should be "Bug #123: WP TITLE" where "WP title is editable" and "Bug" is not a link
- [x] carry query path of url to fullscreen view url so we are properly being redirected when switching back to list or split view
- [x] dropdown actions in "new work package" button are missing so it is not possible to create work packages from the full screen view. (should redirect to create mode in split view)

[1] we initially place the buttons in the top right area. Later @rebellium needs to move them to a bottom bar.
#### Found issues
- [ ] [breadcrumb does nto show project](https://github.com/opf/openproject/pull/3366#issuecomment-140994275)
- [ ] [Can't change view on small resolutions](https://github.com/opf/openproject/pull/3366#issuecomment-140995420)
- [ ] [Request-loop for non-existing work packages](https://github.com/opf/openproject/pull/3366#issuecomment-140998345)
- [ ] [toolbar line-breaking](https://github.com/opf/openproject/pull/3366#issuecomment-141000902)
- [ ] [rendering of anchors](https://github.com/opf/openproject/pull/3366#issuecomment-141015332)
#### Questions
- is it possible to provide a full edit view? how much work would this cause? maybe also in split view mode?
- what happens if someone has no right to view the work package?
#### Out of scope

Create new pull requests for these bugs
- [x] Confirm when deleting an image (#3437)
- [x] Upload image then try to edit the description => error 'conflicting modifications' (#3436)
#### Won't apply
- [x] "new work package" Link in sidebar is linked to the old new view. We should relink it to the new create view (see https://github.com/opf/openproject/pull/3366#issuecomment-135332990)

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://timhabermaas-16364-fullscreen-work-package-view.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://timhabermaas-16364-fullscreen-work-package-view.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by ulferts** [comment]
Thanks for the PR, @timhabermaas. It feels good to finally see it all fit together. Let me know if it is too early to give feedback. 

Here is what I noticed when skimming over the functionality:

The url seems wrong to me:
image
I believe it should be `/work_packages/:id` (please see the work package as well). That should also render the problem of having to fix the routes in the rails world non existent. However, it might introduce new questions when switching between full screen view and details pane/table view especially when having started from the full screen view.

The layout needs some adjustments (especially on large displays:
image

There is no scrollbar on the page leading to content becoming impossible to reach:
image
This is a behaviour inherited from the list view.

The width of the overview content changes depending on whether the "show all" button is enabled or not:
image
vs.
image

The "Relations" tab is not styled properly (the other's look unchanged):
image

It's not possible to create a work package via the view:
image

The settings button is non functional:
image

There are no tests.

**Comment by timhabermaas** [comment]
Thanks for the feedback, @ulferts. It's kinda early, but still appreciated! I wanted to add the `work-in-progress` label to this PR, but ~~don't~~ didn't seem to have the rights to do so. ~~Could someone add it, please?~~ (thanks @NobodysNightmare)

> The url seems wrong to me:

I plan to change it, but need to move some code around first (abstract parent controller depends on the URL being `/projects/:foo/...`).

> That should also render the problem of having to fix the routes in the rails world non existent.

Except for `/work_packages/:id#comment_2`. Anchors will probably break when rendering the page client side.

I'm aware of the design glitches and will fix them at the end. Your tip "This is a behaviour inherited from the list view." is useful, thanks!

> There are no tests.

What kind of tests would you like to see? E.g. there seem to be integration tests using RSpec and some JS integration tests for the work packages index page. Which one should I prefer to modify/extend?

File: frontend/app/routing.js [hunk 1]
```diff
@@ -67,8 +67,51 @@ angular.module('openproject')
...
+    })
+    .state('work-packages.show.watchers', {
+      url: '/watchers',
+      controller: 'DetailsTabWatchersController',
+      templateUrl: '/templates/work_packages/tabs/watchers.html'
+    })
+
     .state('work-packages.list', {
-      url: '',
+      url: '/projects/:projectIdentifier/work_packages?query_id',
```
Comment (timhabermaas) [inline frontend/app/routing.js]: @floriank How bad of an idea is this? Are there other routes besides `/projects/:projectIdentifier/` where the work packages can be nested under?

Comment (floriank) [inline frontend/app/routing.js]: For starters, there is a whole route dedicated to work_packages without a project scope: 

```
/work_packages
```

will show the complete list independent of a project. I will look into this a bit deeper tomorrow.

also, is the `query_id` part necessary? Arbitrary params should be accessible via `$params` or `$stateParams` (I mix those up all the time)

Comment (NobodysNightmare) [inline frontend/app/routing.js]: I can answer to this subpart of the question

> Are there other routes besides /projects/:projectIdentifier/ where the work packages can be nested under?

Nope. We only nest work packages under projects or don't nest them at all. My fantasy also can't come up with other useful nestings, so we should be quite safe from this side ;)

**Comment by timhabermaas** [comment]
@lindenthal @basiszwo Thanks for the feedback (would've preferred it in a different comment instead of editing the first post, though). I'm working through the list right now.

Both of these issues are already present on `dev`:

> Confirm when deleting an image
> Upload image then try to edit the description => error 'conflicting modifications'

I'd prefer to try to fix them in a different branch. Agreed?

**Comment by basiszwo** [comment]
agree

File: frontend/app/routing.js [hunk 2]
```diff
@@ -79,7 +133,7 @@ angular.module('openproject')
       templateUrl: '/templates/work_packages.list.new.html'
     })
     .state('work-packages.list.details', {
-      url: '/{workPackageId:[0-9]+}?query_props',
+      url: '/details/{workPackageId:[0-9]+}?query_props',
```
Comment (ulferts) [inline frontend/app/routing.js]: This change will break existing links people might already have stored outside of the system, .e.g by sending it via mail. Why is it necessary? The former route seemed to be uniq to me as well.

Comment (timhabermaas) [inline frontend/app/routing.js]: It's to differentiate between the fullscreen show view and the split view. If I don't change one of them both have `/work_packages/:id/:tab` as the URL. Changing the show view URL seemed more cumbersome to me.

Old links should still work, but they end up pointing to the fullscreen view instead of the split view now. IMO no big deal: Same information, different layout.

**Comment by basiszwo** [comment]
@timhabermaas I updated the description (reordered the items)

**Comment by basiszwo** [comment]
to get the fullscreen icon you need to set the `icon-fullscreen` css class

**Comment by timhabermaas** [comment]
>  "new work package" Link in sidebar is linked to the old new view. We should relink it to the new create view

Which isn't as trivial as thought: We need to provide a default work package type when linking to the angular new form since there's no option to select the type (URL looks like `/projects/:project_id/work_packages/create_new?type=1`).

**Comment by basiszwo** [comment]
as it is not easy to link to a default type and change it in the form afterwards we are removing the "new" action from the sidebar. This was already in discussion with @lindenthal and now it's going to happen :smile:

**Comment by NobodysNightmare** [comment]
> We need to provide a default work package type when linking to the angular new form

Not sure if this hint applies, but the APIv3 will assume a default type already when a client does not indicate one. E.g. `POST /api/v3/projects/1/work_packages/form` with an empty payload _should_ return a form, where the payload has the default type set.

What I do not know is how/if this applies to the frontend problems you have, but it might be helpful to know.

**Comment by timhabermaas** [comment]
@NobodysNightmare Thanks! I didn't know this. This made the fix pretty easy: 726051efc8477d92b57cf5f59731bf077a2ec035

**Comment by basiszwo** [comment]
see http://clients.zzmedia.net/openproject/workpackages/fullscreen/small/ for an arrangement of the top right icons. The settings (cog) icon on the most upper right will contain some links to the administration area of openproject. The links show in list view don't make sense.
We come up with links soon.

**Comment by basiszwo** [comment]
@timhabermaas the correct icons we need to use are not yet included. Look at https://github.com/opf/openproject/pull/3424/files

we need to replace all 3 icons
- icon-fullscreen-view
- icon-list-view
- icon-split-view

**Comment by ulferts** [comment]
@timhabermaas can you please merge/rebase as it is getting harder and harder to compare and work with this branch. e.g. there are some migrations in dev now that will cause breakages when switching to your branch.

File: frontend/app/routing.js (hunk 2 above)
Comment (ulferts) [inline frontend/app/routing.js]: I think it is not working as you intended, then. 
In the dev and release branches, this is the url for opening a details pane (table + details pane opened with the overview tab): `/projects/:project_id/work_packages/:id/overview`. In this branch, it will yield an empty page. Instead one has to query with `/projects/:project_id/work_packages/details/:id/overview`.

When simply removing the `details` part everything seems to be working again. I am not an expert on the routing but I believe that the state is scoped somehow to only be available scoped under '/projects/:project_id'.

**Comment by timhabermaas** [comment]
@ulferts I rebased.

File: frontend/app/routing.js (hunk 2 above)
Comment (timhabermaas) [inline frontend/app/routing.js]: Ah yeah, missed the nested routes for the show page. I'll fix them once I managed to get the drop-down menu working on the show page.

**Comment by HDinger** [comment]
I started with the styling of the input fields in https://github.com/timhabermaas/openproject/pull/2.

@timhabermaas @basiszwo @lindenthal : could you please have a quick look at the new styling?

**Comment by HDinger** [comment]
This should fix the problem with the scrollbar: https://github.com/timhabermaas/openproject/pull/3

**Comment by HDinger** [comment]
This PR deals with text-overflow values (needed for example for headlines or comments with really long words): https://github.com/timhabermaas/openproject/pull/4

**Comment by HDinger** [comment]
This PR resizes the headlines to full width: https://github.com/timhabermaas/openproject/pull/6

@basiszwo : can you please have a look at it?

**Comment by ulferts** [comment]
I addressed the sub-uri problems in https://github.com/timhabermaas/openproject/pull/8

**Comment by NobodysNightmare** [comment]
One issue that was brought to me: In the full screen view the first activity is displayed as "updated" (instead of "created") and all fields are marked as changed:

pasted image at 2015_09_09 11_05 am

For the split view it works as expected.

Background: This might be due to the lame way I implemented the recognition of the first activity, which was later altered even more by @oliverguenther .

**tl; dr:** You have to tell a user activity whether it `isInitial` explicitly. I am sorry for that. When I first did it, it looked like a quick and "good enough" solution :/

**Comment by basiszwo** [comment]
- "watch" icon shows `true` when watching a WP
- "Watchers" does not show watchers and one can't add watchers. see split view for correct implementation

bildschirmfoto 2015-09-09 um 11 53 20

**Comment by timhabermaas** [comment]
@NobodysNightmare This should be fixed with: 3f5256c9069e1548eff3b7f59720e31359ed11e4 (and 5b3779147545b789c1e602f26dcbce1e1a9abbc7). The `isInitialActivity` wasn't present at all. I love AngularJS for not complaining about such obvious errors. :angry:

**Comment by timhabermaas** [comment]
There's another issue with the split view: Double clicking on an entry doesn't change the shown work package. Removing `reloadOnSearch` in the router fixes this, but also causes the work package list to be reloaded.

**Comment by NobodysNightmare** [comment]
### Breadcrumb does not shows work packages project

#### Reproduction
- open a work package using a direct link (with a project in the URL)
  - e.g. https://community.openproject.org/work_packages/21497

#### Expected (previous behaviour)
- The breadcrumb shows the work packages project
- there is a project specific side menu

image

#### Actual
- breadcrumb shows no project
- no side menu

image

**Comment by NobodysNightmare** [comment]
### Can't change view on small resolutions

#### Reproduction
- use a viewport of 1280px or less width
- open the WP list
- click on a row

#### Expected behaviour
- no inapropriate view mode for my resolution is opened
- in the top right I can choose from available view modes for my resolution
  - list view (always available)
  - split view (IIRC, only available above the break point of 1280px width)
  - fullscreen view (always available)

image

#### Actual
- split view opens
- I have no option to change the view mode

image

**Comment by NobodysNightmare** [comment]
### Request-loop on non-existing work packages

#### Reproduction
- Open the URL for a non-existing (e.g. deleted) work package
  - e.g. community.openproject.org/work_packages/99999 (be careful: better try your local instance!)
- important: do not append the `/activity` part

#### Expected
- some kind of error page, indicating that this is not the work package I am looking for

##### Actual
- a wild request-loop appears, causing some 404 flooding:

image

I guess this has something to do with `work_packages/:id` vs `work_packages/:id/activity`, since the `activity` part in my adress bar began flickering hard...

#### Notes

This also works for work packages I am not allowed to see, but that actually would exist...

**Comment by NobodysNightmare** [comment]
### Subject and toolbar do not always line-break correctly

#### Reproduction
- use a long subject (multiple words)
- look at the fullscreen view at different widths

#### Expected
- line breaking always occurs in a _good looking_™ manner

#### Actual

At different widths the line breaking is sometimes good, sometimes not so...

good:
break-1

not ideal (space on top is enough):
break-2

good (IMO):
break-3

bad (hides part of the subject):
break-4

**Comment by jonasheinrich** [comment]
### Wrong Rendering of anchor

#### Reproduction

Create a work package with a description containing an h1-element.

#### Actual

The anchor is one line too low and thus, under the text of the description. Additionally, it is always displayed.

#### Expected

The anchor is in the correct line. Furthermore, the anchor is only visible when you hover over it. 

bildschirmfoto 2015-09-17 um 10 29 25
